### PR TITLE
Expose a Shipper interface so that clients can create custom Shippers for specialized log processing

### PR DESCRIPTION
**Pull request by corymonroe**
## Goal

I would like to expose this Shipper interface so that I can create a custom backend for any logging pipeline. My specific usecase is that I would like to retain all the special processing this bugsnag client does, but would like to "ship" it to a fluent-bit backend, which handles ingesting, filtering, backpressure, routing, and eventually shipping to another appropriate backend. In this case to bugsnag.

```
bugsnag.Configure(bugsnag.Configuration{
        // Your Bugsnag project API key
        APIKey:          "YOUR API KEY HERE",
        // "fluent-bit shipper, allowing custom log processing outside of the main process."
        Shipper:         NewFluentBitShipper(),
        // more configuration options
    })
```

## Design

This seemed like the most abstract way to add such a feature, without making large sweeping changes to the codebase. This allows anyone to design any sort of backend they please, which can be as simple or as complex as they so choose.

### Added
`Shipper interface`
`httpShipper struct`

### Removed
method `deliver() error` for `payload struct`

### Changed
`payload struct` no longer has a `deliver() error` method
`Configuration struct` now has a default `hTTPShipper`

## Tests

Because I was not adding any new functionality, simply changing the code structure, I have currently relied on the unit tests, although these tests don't appear to have completely passed for quite some time.

### Alternative Approaches

Modifying configuration to take in a generic connection string eg: `tcp:localhost:7777` etc. This was not nearly as flexible as the above route.

### Outstanding Questions

The name `Shipper` was probably what I spent the most time on here... I am open to other suggestions on that name, as well as the rest of my code!

## Review

For the submitter, initial self-review:

- [x] Commented on code changes inline explain the reasoning behind the approach
- [ ] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [x] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: bugsnag.go [hunk 1]
```diff
@@ -139,6 +139,7 @@ func init() {
 		ProjectPackages:     []string{"main*"},
 		NotifyReleaseStages: nil,
 		Logger:              log.New(os.Stdout, log.Prefix(), log.Flags()),
+		Shipper:             NewHTTPShipper(),
```
Comment (cmonroe13) [inline bugsnag.go]: Not entirely sure I need to make this method public.

File: configuration.go [hunk 2]
```diff
@@ -68,6 +68,10 @@ type Configuration struct {
 	Logger interface {
 		Printf(format string, v ...interface{}) // limited to the functions used
 	}
+	// Shipper is the expected interface for sending payload data to a backend
+	Shipper interface {
+		Deliver(p *payload) error
```
Comment (cmonroe13) [inline configuration.go]: This could also be private, I believe.

File: shipper_http.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+)
+
+// NewHTTPShipper creates a shipper to send data over http
+func NewHTTPShipper() Shipper {
```
Comment (cmonroe13) [inline shipper_http.go]: Not entirely sure this method should remain public.

File: shipper_http.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	return &httpShipper{}
+}
+
+type httpShipper struct {
+}
+
+func (s *httpShipper) Deliver(p *payload) error {
```
Comment (cmonroe13) [inline shipper_http.go]: This method, Deliver(p *payload) error could also be private.

**Comment by cmonroe13** [comment]
Anything I should be looking to add before I can get some initial impressions and/or review? Would love to get this merged and/or hear your thoughts. We can always buffer the notify calls in the app level, but would love to ship them to a custom backend that handles filtering etc.

**Comment by cmonroe13** [comment]
Another alternative PR would be to JUST make `payload` public and `newEvent` method from `Event` public. That would allow us to just use what we needed, which is the `Configuration`, `Event`, and `Payload`. We could then build our own client that wrapped these three things.

**Comment by kattrali** [comment]
Hello! I apologize for the delay. Thank you for the contribution, @cmonroe13. 

Making the delivery mechanism replaceable makes a lot of sense, and in fact we have a generic interface for this on other platforms, like [python](https://github.com/bugsnag/bugsnag-python/blob/9ed199557cecb622d59af8bccefae7fa844c4678/bugsnag/delivery.py#L45) and [ruby](https://github.com/bugsnag/bugsnag-ruby/blob/56f387ef36c157a2b6fbf7d9e6807d21d1e62393/lib/bugsnag/delivery.rb#L2), where people can register a custom mechanism at the configuration level. If we change this for Go, we need to use similar terminology (which helps people using multiple platforms). Making this change will require a bit of thought on our part, to ensure we are being consistent but also forward-thinking and Go-like.

This looks like the correct general approach, though we'll need some time to review.
